README.md
# Snapfolio - VSCode Extension

## Description

Snapfolio is a Visual Studio Code extension that seamlessly integrates with Notion, allowing you to enhance your workflow by customizing code snippets, adding text and code to Notion pages, creating new Notion pages, and exploring your content.

## Features

- **Authenticate with Notion:** Connect your Notion account to Snapfolio.
- **Customize Snippets:** Tailor your code snippets with themes, font size, and background color.
- **Add Text and Code to Notion Pages:** Enhance your Notion pages directly from VSCode.
- **Create Notion Pages:** Quickly create new Notion pages with a specified title.
- **Explorer View:** Navigate and manage your Snapfolio content directly within VSCode.

## Requirements

- VSCode version 1.84.0 or higher.
- Notion account.

## Installation

1. Launch VS Code.
2. Go to Extensions.
3. Search for "Snapfolio" and click Install.

## Local Installation

1. Clone the repository.

    ```bash
    git clone https://github.com/Jasleen8801/snapfolio.git
    ```

2. Clone the other repository.

    ```bash
    git clone https://github.com/Jasleen8801/Snapfolio-API.git
    ```
  
3. Setup the backend repo as given in instructions [here](https://github.com/Jasleen8801/Snapfolio-API/blob/main/README.md).

4. Open the project in VSCode.

5. Install dependencies.

    ```bash
    npm install
    ```
  
6. Build the project.

    ```bash
    npm run build
    ```

7. Run the project.

    ```bash
    npm run watch
    ```

8. Open a new VSCode window and open the project.

9. Create a .env file in the root directory of the project and add the following:

    ```bash
    AUTHORIZATION_URL=<YOUR_AUTH_URL_AS_IN_BACKEND_API>
    ```

10. Press `F5` to launch the extension in a new window.

11. Authenticate Snapfolio using the command palette (`Ctrl+Shift+P` or `Cmd+Shift+P`) and select "Snapfolio: Authenticate."

**Note**: The backend API assumes to be run on port 3000. If you want to change the port, make sure to change the port in the backend API as well. 

## Getting Started

1. Open your project in VSCode.
2. Authenticate Snapfolio using the command palette (`Ctrl+Shift+P` or `Cmd+Shift+P`) and select "Snapfolio: Authenticate."
3. Customize snippets, add text to Notion, create Notion pages, and explore content using the available commands.

## Commands

- `Snapfolio: Authenticate`: Authenticate Snapfolio with your Notion account.
- `Snapfolio: Customize Snippet`: Customize code snippet settings.
- `Snapfolio: Add Text to Notion`: Add text to your Notion pages.
- `Snapfolio: Add Code to Notion`: Add code (programming + output) to your Notion pages.
- `Snapfolio: Create Notion Page`: Create a new Notion page.
- `Snapfolio: Explorer View`: Open the Snapfolio Explorer View.

## Configuration

- `snapfolio.theme`: Set the default theme for code snippets.
- `snapfolio.fontSize`: Set the default font size for code snippets.
- `snapfolio.backgroundColor`: Set the default background color for code snippets.
- `snapfolio.botId`: Store the bot ID after authentication.
- `snapfolio.pageId`: Store the current page ID.

## Backend Node-Express API

Snapfolio relies on a Node-Express backend API for handling Notion interactions. Explore the backend repository at [Backend Repository](https://github.com/Jasleen8801/Snapfolio-API).

## Feedback

If you encounter any issues or have suggestions, please [open an issue](https://github.com/Jasleen8801/snapfolio/issues) on GitHub.

## Contributing

Feel free to contribute to the development of Snapfolio. Fork the repository, make your changes, and submit a pull request.


CHANGELOG.md
# Change Log

All notable changes to the "snapfolio" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/addCode.ts
import * as vscode from 'vscode';
import axios from 'axios';

export function activate(context: vscode.ExtensionContext) {
  const API_URI: string = "http://localhost:3000";

  let addCodeDisposable = vscode.commands.registerCommand('snapfolio.addCode', async () => {
    const bot_id = vscode.workspace.getConfiguration('snapfolio').get('botId');
    const theme = vscode.workspace.getConfiguration('snapfolio').get('theme');
    const fontSize = vscode.workspace.getConfiguration('snapfolio').get('fontSize');
    const backgroundColor = vscode.workspace.getConfiguration('snapfolio').get('backgroundColor');
    const textEditor = vscode.window.activeTextEditor;
    const page_id = vscode.workspace.getConfiguration('snapfolio').get('pageId');
    let flag: boolean = false;
    if(textEditor){
      const selectedText = textEditor.document.getText(textEditor.selection);
      if(selectedText){
        const subheading = await vscode.window.showInputBox({
          prompt: "Enter a subheading for the code block",
          placeHolder: "Enter a subheading",
        }) || '';
        const description = await vscode.window.showInputBox({
          prompt: "Enter a description for the code block",
          placeHolder: "Enter a description",
        }) || '';
        const ans: string = await vscode.window.showInputBox({
          prompt: "Do you want to add the output to the document?",
          placeHolder: "Enter (y/n)",
        }) || 'n';
        // console.log(ans);
        if(ans === 'y' || ans === 'Y'){
          flag = true;
          const config = vscode.workspace.getConfiguration();
          config.update('terminal.integrated.copyOnSelection', true, vscode.ConfigurationTarget.Global);
          const selectedOutput = await vscode.env.clipboard.readText();

          const output = await vscode.window.showInputBox({
            prompt: "Check the output selected in the terminal",
            value: selectedOutput,
          });
          console.log(output);

          const response = await axios.post(`${API_URI}/notion/appendCode`, {
            bot_id: bot_id,
            flag: flag,
            code: selectedText,
            subheading: subheading,
            description: description,
            theme: theme,
            fontSize: fontSize,
            backgroundColor: backgroundColor,
            output: selectedOutput,
            page_id: page_id
          });
          console.log(response.data);
        } else if(ans === 'n'){
          const response = await axios.post(`${API_URI}/notion/appendcode`, {
            bot_id: bot_id,
            flag: flag,
            code: selectedText,
            subheading: subheading,
            description: description,
            theme: theme,
            fontSize: fontSize,
            backgroundColor: backgroundColor,
            page_id: page_id
          });
          console.log(response);
        }
      }
    }
  });

  context.subscriptions.push(addCodeDisposable);
}

export function deactivate() { }

src/addText.ts
import * as vscode from 'vscode';
import axios from 'axios';

export function activate(context: vscode.ExtensionContext) {
	const API_URI: string = "http://localhost:3000";
  
  let addTextDisposable = vscode.commands.registerCommand('snapfolio.addText', async () => {
		const bot_id = vscode.workspace.getConfiguration('snapfolio').get('botId');
		// console.log(bot_id);

		if (bot_id) {
			const panel = vscode.window.createWebviewPanel(
				'addText',
				'Notion Page Customization',
				vscode.ViewColumn.One,
				{
					enableScripts: true
				}
			);

			panel.webview.html = html_content;
			
			panel.webview.onDidReceiveMessage(async (message) => {
				if(message.command === 'customize'){
					const { title, description } = message;
					console.log(title, description);
					const pageId = vscode.workspace.getConfiguration('snapfolio').get('pageId');
					const response = await axios.post(`${API_URI}/notion/appendText`, {
						bot_id: bot_id,
						boldText: title,
						text: description,
						page_id: pageId
					});
					console.log(response);
					if(response.status === 200){
						vscode.window.showInformationMessage('Text Added Successfully!');
						panel.dispose();
					} else {
						vscode.window.showErrorMessage('Text Addition failed!');
					}
				} else {
					console.log(message);
				}
			});
		}
	});
  
  context.subscriptions.push(addTextDisposable);
}

export function deactivate() { }

let html_content = `
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Add Text</title>
  <style>
    body {
      font-family: 'Arial', sans-serif;
      margin: 20px;
      text-align: center;
    }

    h1 {
      color: #333;
    }

    label {
      display: block;
      margin-top: 10px;
      margin-bottom: 5px;
      color: #555;
    }

    input {
      padding: 8px;
      margin-bottom: 10px;
      box-sizing: border-box;
    }

    button {
      padding: 10px 20px;
      font-size: 16px;
      background-color: #4CAF50;
      color: white;
      border: none;
      border-radius: 5px;
      cursor: pointer;
    }

    button:hover {
      background-color: #45a049;
    }
  </style>
</head>
<body>
  <h1>Add Text to Notion Page</h1>
  <div>
    <label for="title">Subtitle: </label>
    <input id="title" type="text" placeholder="(in bold)">
  </div>
  <div>
    <label for="description">Text: </label>
    <input id="description" type="text" placeholder="Description">
  </div>
  <button id="customizeButton">Submit</button>

  <script>
    const vscode = acquireVsCodeApi();
    const customizeButton = document.getElementById('customizeButton');

    customizeButton.addEventListener('click', () => {
      const title = document.getElementById('title').value;
      const description = document.getElementById('description').value;

      vscode.postMessage({
        command: 'customize',
        title: title,
        description: description
      });
    });
  </script>
</body>
</html>`;


src/extension.ts
import * as vscode from 'vscode';

import { activate as authenticateActivate, deactivate as authenticateDeactivate } from './authenticate';
import { activate as addTextActivate, deactivate as addTextDeactivate } from './addText';
import { activate as addCodeActivate, deactivate as addCodeDeactivate } from './addCode';
import { activate as customizeSnippetActivate, deactivate as customizeSnippetDectivate } from './customizeSnippet';
import { activate as createPageActivate, deactivate as createPageDeactivate } from './createPage';
import { activate as explorerViewActivate, deactivate as explorerViewDeactivate } from './explorerView';

export function activate(context: vscode.ExtensionContext) {
	authenticateActivate(context);
	addTextActivate(context);
	addCodeActivate(context);
	customizeSnippetActivate(context);
	createPageActivate(context);
	explorerViewActivate(context);
}

export function deactivate() { }


src/createPage.ts
import * as vscode from 'vscode';
import axios from 'axios';

export function activate(context: vscode.ExtensionContext){
  const API_URI: string = "http://localhost:3000";

  let createPageDisposable = vscode.commands.registerCommand('snapfolio.createPage', async() => {
    const bot_id = vscode.workspace.getConfiguration('snapfolio').get('botId');
    if(bot_id){
      const panel = vscode.window.createWebviewPanel(
        'createPage',
        'Create a Notion Page',
        vscode.ViewColumn.One,
        {
          enableScripts: true
        }
      );

      panel.webview.html = html_content;

      panel.webview.onDidReceiveMessage(async (message) => {
        if(message.command === 'create'){
          const { title } = message;
          console.log(title);
          const response = await axios.post(`${API_URI}/notion/createPage`, {
            bot_id: bot_id,
            title: title
          });
          const { page_id } = response.data;
          console.log(page_id);
          panel.dispose();
          vscode.window.showInformationMessage(response.data.message);
        }
      });
    }
  });

  context.subscriptions.push(createPageDisposable);
}

export function deactivate() {}

let html_content =  `
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Create a Notion Page</title>
  <style>
    body {
      font-family: 'Arial', sans-serif;
      margin: 20px;
      text-align: center;
    }

    h1 {
      color: #333;
    }

    label {
      display: block;
      margin-top: 10px;
      margin-bottom: 5px;
      color: #555;
    }

    input {
      padding: 8px;
      margin-bottom: 10px;
      box-sizing: border-box;
    }

    button {
      padding: 10px 20px;
      font-size: 16px;
      background-color: #4CAF50;
      color: white;
      border: none;
      border-radius: 5px;
      cursor: pointer;
    }

    button:hover {
      background-color: #45a049;
    }
  </style>
</head>
<body>
  <h1>Create a Notion Page</h1>
  <div>
    <label for="title">Page Title: </label>
    <input id="title" type="text" placeholder="Page Title">
  </div>
  <button id="createButton">Create</button>

  <script>
    const vscode = acquireVsCodeApi();
    const createButton = document.getElementById('createButton');

    createButton.addEventListener('click', () => {
      const title = document.getElementById('title').value;
      vscode.postMessage({
        command: 'create',
        title: title
      });
    });
  </script>
</body>
</html>`;


src/authenticate.ts
import * as vscode from 'vscode';
import * as dotenv from 'dotenv';
import axios from 'axios';

export function activate(context: vscode.ExtensionContext) {
  dotenv.config();
	const AUTH_URL: string = process.env.AUTHORIZATION_URL || '';
	const API_URI: string = "http://localhost:3000";

  let authDisposable = vscode.commands.registerCommand('snapfolio.authenticate', async () => {
		const targetUri = vscode.Uri.parse(AUTH_URL);
		await vscode.env.openExternal(targetUri);

		const panel = vscode.window.createWebviewPanel(
			'snapfolio',
			'Snapfolio',
			vscode.ViewColumn.One,
			{
				enableScripts: true
			}
		);

		panel.webview.html = html_content;

		panel.webview.onDidReceiveMessage(async (message) => {
			if (message.command === 'authenticate') {
				const input: string = await vscode.window.showInputBox({
					prompt: 'Enter your final Uri you copied',
					placeHolder: 'Final URI'
				}) || '';
				// console.log(input);
				if (input !== '') {
					const url = new URL(input);
					const code = url.searchParams.get('code');
					// console.log(code);
					if (code) {
						try {
							const response = await axios.post(`${API_URI}/auth/redirect_uri?code=${code}`);
							// console.log(response);
							if (response.status === 200) {
								const { bot_id } = response.data;
								const config = vscode.workspace.getConfiguration('snapfolio');
								await config.update('botId', bot_id, vscode.ConfigurationTarget.Global);
								vscode.window.showInformationMessage('Authentication successful!');
								panel.dispose();
							} else {
								vscode.window.showErrorMessage('Authentication failed!');
							}
						} catch (error) {
							console.log(error);
						}
					}
				}

			}
		});
	});

  context.subscriptions.push(authDisposable);
}

export function deactivate() { }

let html_content = `<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1.0">
	<title>Snapfolio Authentication</title>
	<style>
		body {
			font-family: 'Arial', sans-serif;
			margin: 20px;
			text-align: center;
		}

		h1 {
			color: #333;
		}

		p {
			color: #666;
		}

		button {
			padding: 10px 20px;
			font-size: 16px;
			background-color: #4CAF50;
			color: white;
			border: none;
			border-radius: 5px;
			cursor: pointer;
		}

		button:hover {
			background-color: #45a049;
		}
	</style>
</head>
<body>
	<h1>Snapfolio Authentication</h1>
	<p>Click the button below to authenticate.</p>
	<button id="authenticate">Authenticate</button>

	<script>
		const vscode = acquireVsCodeApi();
		const button = document.getElementById('authenticate');

		button.addEventListener('click', () => {
			vscode.postMessage({
				command: 'authenticate'
			});
		});
	</script>
</body>
</html>`;



src/explorerView.ts
import * as vscode from 'vscode';
import axios from 'axios';

export function activate(context: vscode.ExtensionContext) {
  const API_URI: string = "http://localhost:3000";

  const statusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 100);
  statusBarItem.text = "$(book) Snapfolio";
  statusBarItem.command = 'snapfolio.explorerView';

  let explorerViewDisposable = vscode.commands.registerCommand('snapfolio.explorerView', async () => {
    try {
      const bot_id = vscode.workspace.getConfiguration('snapfolio').get('botId');
      const response = await axios.get(`${API_URI}/notion/getPages`, {
        params: {
          bot_id: bot_id
        }
      });

      const pages = response.data.pages;
      const pageTitles = pages.map((page: any) => page.title);
      const selectedPage = await vscode.window.showQuickPick(pageTitles);
      const index = pageTitles.indexOf(selectedPage || '') || 0;

      statusBarItem.text = `$(book) Snapfolio - ${selectedPage}`;

      await vscode.workspace.getConfiguration('snapfolio').update('pageId', pages[index].id, vscode.ConfigurationTarget.Global);
      console.log(pages[index].id);
      vscode.window.showInformationMessage(`Selected Page: ${selectedPage}`);
    } catch (error) {
      vscode.window.showErrorMessage(`Please choose a page first!`);
    }
  });

  context.subscriptions.push(explorerViewDisposable);
  context.subscriptions.push(statusBarItem);

  statusBarItem.show();
}

export function deactivate() { }


src/customizeSnippet.ts
import * as vscode from 'vscode';

export function activate(context: vscode.ExtensionContext) {
  let customizeSnippetDisposable = vscode.commands.registerCommand('snapfolio.customizeSnippet', async () => {
    const panel = vscode.window.createWebviewPanel(
      'customizeSnippet',
      'Customize Snippet',
      vscode.ViewColumn.One,
      {
        enableScripts: true
      }
    );
    
    panel.webview.html = html_content;
    
    panel.webview.onDidReceiveMessage(async (message) => {
      if(message.command === 'customizeSnippet'){
        const { theme, fontSize, backgroundColor } = message;
        console.log(theme, fontSize, backgroundColor);
        const config = vscode.workspace.getConfiguration();
        config.update('snapfolio.theme', theme, vscode.ConfigurationTarget.Global);
        config.update('snapfolio.fontSize', fontSize, vscode.ConfigurationTarget.Global);
        config.update('snapfolio.backgroundColor', backgroundColor, vscode.ConfigurationTarget.Global);
        panel.dispose();
      }
    });
  });

  context.subscriptions.push(customizeSnippetDisposable);
}

export function deactivate() { }

let html_content = `
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Customize Snippet Settings</title>
  <style>
    body {
      font-family: 'Arial', sans-serif;
      margin: 20px;
      text-align: center;
    }

    h1 {
      color: #333;
    }

    label {
      display: block;
      margin-top: 10px;
      margin-bottom: 5px;
      color: #555;
    }

    input, select {
      padding: 8px;
      margin-bottom: 10px;
      box-sizing: border-box;
    }

    button {
      padding: 10px 20px;
      font-size: 16px;
      background-color: #4CAF50;
      color: white;
      border: none;
      border-radius: 5px;
      cursor: pointer;
    }

    button:hover {
      background-color: #45a049;
    }

    #feedback {
      margin-top: 15px;
      font-size: 14px;
      color: #777;
    }
  </style>
</head>
<body>
  <h1>Customize Snippet Settings</h1>
  <div>
    <label for="theme">Theme: </label>
    <select id="theme">
      <option value="3024-night">3024 Night</option>
      <option value="a11y-dark">A11y Dark</option>
      <option value="blackboard">Blackboard</option>
      <option value="base16-dark">Base16 Dark</option>
      <option value="cobalt">Cobalt</option>
      <option value="darcula">Darcula</option>
      <option value="dracula">Dracula</option>
      <option value="duotone-dark">Duotone Dark</option>
      <option value="gruvbox-dark">Gruvbox Dark</option>
      <option value="hopscotch">Hopscotch</option>
      <option value="material">Material</option>
      <option value="material-darker">Material Darker</option>
      <option value="material-palenight">Material Palenight</option>
      <option value="nord">Nord</option>
      <option value="oceanic-next">Oceanic Next</option>
      <option value="panda-syntax">Panda Syntax</option>
      <option value="paraiso-dark">Paraiso Dark</option>
      <option value="seti">Seti</option>
      <option value="shades-of-purple">Shades of Purple</option>
      <option value="solarized">Solarized</option>
      <option value="synthwave-84">SynthWave '84</option>
      <option value="twilight">Twilight</option>
      <option value="verminal">Verminal</option>
      <option value="vscode">Visual Studio Code</option>
      <option value="yeti">Yeti</option>
      <option value="zenburn">Zenburn</option>
    </select>
  </div>
  <div>
    <label for="fontSize">Font Size: </label>
    <input id="fontSize" type="number" placeholder="Enter font size">
  </div>
  <div>
    <label for="backgroundColor">Background Color: </label>
    <input id="backgroundColor" type="color" placeholder="Choose background color">
  </div>
  <button id="customizeButton">Submit</button>
  <div id="feedback"></div>

  <script>
    const vscode = acquireVsCodeApi();
    const customizeButton = document.getElementById('customizeButton');
    const feedbackElement = document.getElementById('feedback');

    customizeButton.addEventListener('click', () => {
      const theme = document.getElementById('theme').value;
      const fontSize = document.getElementById('fontSize').value;
      const backgroundColor = document.getElementById('backgroundColor').value;

      vscode.postMessage({
        command: 'customizeSnippet',
        theme: theme,
        fontSize: fontSize,
        backgroundColor: backgroundColor
      });
    });

    window.addEventListener('message', (event) => {
      if (event.data.command === 'feedback') {
        const { success, message } = event.data;
        feedbackElement.textContent = message;
        feedbackElement.style.color = success ? '#4CAF50' : '#FF6347'; // Green for success, red for error
      }
    });
  </script>
</body>
</html>`;
